feat: create table component based on ant design Table

eric-briscoe opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->
Adds a new Table component based on Ant Design as part of movement to a Superset Design system
### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
To view the table component and see the documentation:
https://75b68f5--62b4d4c211d03654c1992245.chromatic.com/?path=/story/design-system-components-table-overview--page (manually updating to Chromatic)

To run local:
1.  checkout this branch
2. open a terminal at the superset/superset-frontend directory and run `npm ci`
3. in the same terminal run `npm run storybook`
4. Open http://localhost:6006/?path=/docs/design-system-components-table-overview--page in your browser

This provides the documentation to use the new Table component as well as interactive stories to see the behavior of the various props

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

eric-briscoe committed:
ant design Table component work in progress

Very rough starting point for new table component

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Work in progress, enabled resizeable columns in Table, added draft design system overview mdx, and Table overview mdx

eric-briscoe committed:
Update DesignSystem.stories.mdx

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Updates documentation, adds tests, separates resizable column and drag drop code into separate class

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Updates to markdown overview docs for table and additional Storybook stories

eric-briscoe committed:
Adds numeric cell renderer, updated buton cell renderer, adds documentation, fixes re-orderable drag drop columns issue

eric-briscoe committed:
Cell renderer work in progress

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Fixes Lint and TS errors

eric-briscoe committed:
Improving folder structure, adding additional tests

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Removes unnecessary code causing an import error

pkdotson commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+  /**
+   * Enables setting the text displayed in various components and tooltips within the Table UI.
+   */
+  locale?: Locale;
```
nit: maybe make this LacaleText?

eric-briscoe replied:
@pkdotson agree the name could be better with better clarity, but in this case we are mirroring some of the props names from Ant Design.  `locale` is how the prop is named for ant d and keeping name matching reduces amount of prop name redefining between the Table component in the internal ant d table instance.

AAfghahi commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+          this.RESIZE_INDICATOR_THRESHOLD,
+        )
+      ) {
+        // @ts-ignore
```
what was happening that required a ts-ignore here?

eric-briscoe replied:
@arash TypeScript does not like the image import, bit this is valid for webpack to import the resource

AAfghahi commented on superset-frontend/src/components/Table/cell-renderers/ButtonCell/ButtonCell.stories.tsx:
```diff
@@ -0,0 +1,63 @@
+} as ComponentMeta<typeof ButtonCell>;
+
+// eslint-disable-next-line no-alert
+const clickHandler = () => alert(`I was Clicked`);
```
This should be deleted right?

AAfghahi commented on superset-frontend/src/components/Loading/index.tsx:
```diff
@@ -20,6 +20,8 @@
 import React from 'react';
 import { styled } from '@superset-ui/core';
 import cls from 'classnames';
+// @ts-ignore
```
what ts issue is this ignoring?

eric-briscoe replied:
@arash TypeScript does not like the image import, bit this is valid for webpack to import the resource

AAfghahi commented:
Approved but have a few questions!

lyndsiWilliams commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/index.tsx:
```diff
@@ -0,0 +1,144 @@
+   * Optional variable that can contain data relevant to the menu item that you
+   * want easy access to in the callback function for the menu
+   */
+  payload?: any;
```
Can this `any` be defined?

eric-briscoe replied:
@lyndsiWilliams This is intentionally open ended so that the user of `payload` can pass a value of any type.  This defined by the developer creating an action cell and is passed back to them when an action is invoked.  The value here is never used internally and we do not need to know its shape.  Because the developer can pass any value they want, we cannot provide a type for this attribute

lyndsiWilliams commented on superset-frontend/src/components/Table/cell-renderers/NumericCell/NumericCell.test.tsx:
```diff
@@ -0,0 +1,50 @@
+  expect(screen.getByText('5 678 943,00 €')).toBeInTheDocument();
+});
+
+test('renders with english US locale and USD currency format', async () => {
```
```suggestion
test('renders with English US locale and USD currency format', () => {
```
Same as above, no async in this test.

eric-briscoe replied:
updating in next commit

lyndsiWilliams commented on superset-frontend/src/components/Table/cell-renderers/NumericCell/index.tsx:
```diff
@@ -0,0 +1,418 @@
+  let displayValue = value?.toString();
+  try {
+    displayValue = value?.toLocaleString?.(locale, options);
+  } catch (e: any) {
```
Can this `any` be defined?

eric-briscoe replied:
Removing typing here as catch errors are not typically typed

lyndsiWilliams commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+   */
+  sticky?: boolean;
+  /**
+   * Controls if columns are re-sizeable by user.
```
```suggestion
   * Controls if columns are resizable by user.
```

lyndsiWilliams commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+  locale?: Locale;
+  /**
+   * Restricts the visible height of the table and allows for internal scrolling within the table
+   * when the number of rows exceed the visible space.
```
```suggestion
   * when the number of rows exceeds the visible space.
```

lyndsiWilliams commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+// This accounts for the tables header and pagination if user gives table instance a height. this is a temp solution
+const HEIGHT_OFFSET = 108;
+
+const StyledTable: StyledComponent<any> = styled(AntTable)<any>`
```
Can these `any`s be defined?

lyndsiWilliams commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+
+const defaultRowSelection: React.Key[] = [];
+// This accounts for the tables header and pagination if user gives table instance a height. this is a temp solution
+const HEIGHT_OFFSET = 108;
```
`HEIGHT_OFFSET` should use the Superset theme's gridvalue since it's a set height.

eric-briscoe replied:
This is odd case where it has to do with internals of an antd component.  Theme gridValue should not be used here as it could have undesired side effect of changing the value from 108 to something else (if gridValue is adjusted in theme) while gridValue will have no affect on the internals of the antd Table

lyndsiWilliams commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+  }
+
+  th.ant-table-cell {
+    font-weight: 600;
```
```suggestion
    font-weight: ${theme.typography.weights.bold};
```

lyndsiWilliams commented on superset-frontend/src/components/Table/sorters.test.ts:
```diff
@@ -0,0 +1,103 @@
+  expect(numericalSort('name', undefined, undefined)).toBe(NaN);
+  // @ts-ignore
+  expect(numericalSort('name', null, null)).toBe(NaN);
+  // incorect non-object values
+  // @ts-ignore
+  expect(numericalSort('name', 3, [])).toBe(NaN);
+  // incorect object values without specificed key
+  expect(numericalSort('name', {}, {})).toBe(NaN);
+  // Object as value for naemw when it should be a string
```
```suggestion
  // incorrect non-object values
  // @ts-ignore
  expect(numericalSort('name', 3, [])).toBe(NaN);
  // incorrect object values without specified key
  expect(numericalSort('name', {}, {})).toBe(NaN);
  // Object as value for name when it should be a string
```

lyndsiWilliams commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+
+  reorderable: boolean;
+
+  derivedColumns: Array<any>;
```
Can this `any` be defined?

eric-briscoe replied:
updated to remove any in next commit

lyndsiWilliams commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+
+  constructor(
+    tableRef: HTMLTableElement,
+    derivedColumns: Array<any>,
```
Can this `any` be defined?

lyndsiWilliams commented on superset-frontend/src/components/Table/utils/utils.ts:
```diff
@@ -0,0 +1,40 @@
+ */
+
+/**
+ * Method to check if a number is within inclusive range between a maximum values minus a threshold
```
```suggestion
 * Method to check if a number is within inclusive range between a maximum value minus a threshold
```

lyndsiWilliams reviewed (commented):
I went through and found some things/had some questions. I also noticed that a storybook log was pushed up to `superset-frontend/build-storybook.log`. I don't see storybook logs elsewhere in the codebase, was this log meant to be part of this PR?

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/sorters.test.ts

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/utils/utils.ts

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+        );
+      }
+      if (resizable) {
+        interactiveTableUtils?.current?.initializeResizeableColumns(
```
Fixing this in next commit as it also requires change in the interactiveTableUtils file

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx:
````diff
@@ -0,0 +1,70 @@
+```
+import ActionCell from './index';
+
+// NOTE: include the other columns you
````
removing comment in next comit

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Lyndsi Kay Williams <55605634+lyndsiWilliams@users.noreply.github.com>

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/NumericCell/NumericCell.test.tsx:
```diff
@@ -0,0 +1,50 @@
+import { render, screen } from 'spec/helpers/testing-library';
+import NumericCell, { CurrencyCode, LocaleCode, Style } from './index';
+
+test('renders with french locale and Euro currency format', async () => {
```
Updating in next commit

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.test.tsx:
```diff
@@ -0,0 +1,49 @@
+  exampleMenuOptions[0].onClick = clickHandler;
+  render(<ActionCell menuOptions={exampleMenuOptions} row={exampleRow} />);
+  // Open the menu
+  await userEvent.click(screen.getByTestId('dropdown-trigger'));
```
updating in next commit

eric-briscoe commented on superset-frontend/src/components/Table/Table.test.tsx:
```diff
@@ -0,0 +1,79 @@
+  render(
+    <Table size={TableSize.MIDDLE} columns={testColumns} data={testData} />,
+  );
+  expect(screen.getByText(testColumns[0].title as string)).toBeInTheDocument();
```
changed in next commit

eric-briscoe committed:
Updates addressing PR comments for ACT Test errors, removal to type: any in some places, and fixing typos

eric-briscoe committed:
Merge branch 'ericbriscoe/sc-51384/create-table-component-based-on-antd' of ssh://github.com/preset-io/superset into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Remove addition to global style for action cell shadows

eric-briscoe committed:
Merge branch 'master' into ericbriscoe/sc-51384/create-table-component-based-on-antd

eric-briscoe committed:
Removing storybook log file which should not have bend comitted

michael-s-molina commented on superset-frontend/src/components/Table/cell-renderers/NumericCell/index.tsx:
```diff
@@ -0,0 +1,419 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
```
```suggestion
import React from 'react';
import { logging } from '@superset-ui/core';
```

michael-s-molina commented on superset-frontend/src/components/Table/cell-renderers/NumericCell/index.tsx:
```diff
@@ -0,0 +1,419 @@
+  try {
+    displayValue = value?.toLocaleString?.(locale, options);
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error(e);
```
```suggestion
    logging.error(e);
```

michael-s-molina commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
```
We generally don't leave a blank line after the license. It's not a blocker for me, but if you choose to remove it to keep the codebase consistent then you can update the other files.

eric-briscoe replied:
fixing in next commit

michael-s-molina commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+const HEIGHT_OFFSET = 108;
+
+const StyledTable: StyledComponent<any> = styled(AntTable)<any>`
+  ${({ theme, height }) => `
+  .ant-table-body {
+    overflow: scroll;
+    height: ${height ? `${height - HEIGHT_OFFSET}px` : undefined};
+  }
+
+  th.ant-table-cell {
+    font-weight: ${theme.typography.weights.bold};
+    color: ${theme.colors.grayscale.dark1};
+    user-select: none;
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  }
+
+  .ant-pagination-item-active {
+    border-color: ${theme.colors.primary.base};
+  }
+  `}
```
```suggestion
  ${({ theme, height }) => `
    .ant-table-body {
      overflow: scroll;
      height: ${height ? `${height - HEIGHT_OFFSET}px` : undefined};
    }

    th.ant-table-cell {
      font-weight: ${theme.typography.weights.bold};
      color: ${theme.colors.grayscale.dark1};
      user-select: none;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ant-pagination-item-active {
      border-color: ${theme.colors.primary.base};
    }
  `}
```

michael-s-molina commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+import React, { useState, useEffect, useRef, ReactElement } from 'react';
+import { Table as AntTable, ConfigProvider } from 'antd';
+import type { ColumnsType, TableProps as AntTableProps } from 'antd/es/table';
+import { t, useTheme } from '@superset-ui/core';
```
```suggestion
import { t, useTheme, logging } from '@superset-ui/core';
```

michael-s-molina commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+  // Log use of experimental features
+  useEffect(() => {
+    if (reorderable === true) {
+      // eslint-disable-next-line no-console
+      console.warn(
```
```suggestion
      logging.warn(
```

michael-s-molina commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+      );
+    }
+    if (resizable === true) {
+      // eslint-disable-next-line no-console
+      console.warn(
```
```suggestion
      logging.warn(
```

michael-s-molina commented on superset-frontend/src/components/Table/index.tsx:
```diff
@@ -0,0 +1,320 @@
+      interactiveTableUtils?.current?.clearListeners?.();
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [wrapperRef, reorderable, resizable]);
```
Can you add the `initializeInteractiveTable` dependency and wrap the function with `useCallback`?

eric-briscoe replied:
@michael-s-molina I made these changes but it actually caused the resizing to stop working correctly:

`  const initializeInteractiveTable = useCallback(() => {
    if (interactiveTableUtils.current) {
      interactiveTableUtils.current?.clearListeners();
    }
    const table = wrapperRef.current?.getElementsByTagName('table')[0];
    if (table) {
      interactiveTableUtils.current = new InteractiveTableUtils(
        table,
        derivedColumns,
        setDerivedColumns,
      );
      if (reorderable) {
        interactiveTableUtils?.current?.initializeDragDropColumns(
          reorderable,
          table,
        );
      }
      if (resizable) {
        interactiveTableUtils?.current?.initializeResizableColumns(
          resizable,
          table,
        );
      }
    }
  }, [derivedColumns, reorderable, resizable]);

  // Log use of experimental features
  useEffect(() => {
    if (reorderable === true) {
      logging.warn(
        'EXPERIMENTAL FEATURE ENABLED: The "reorderable" prop of Table is experimental and NOT recommended for use in production deployments.',
      );
    }
    if (resizable === true) {
      logging.warn(
        'EXPERIMENTAL FEATURE ENABLED: The "resizable" prop of Table is experimental and NOT recommended for use in production deployments.',
      );
    }
  }, [reorderable, resizable]);

  useEffect(() => {
    initializeInteractiveTable();
    return () => {
      interactiveTableUtils?.current?.clearListeners?.();
    };
  }, [wrapperRef, reorderable, resizable, initializeInteractiveTable]);`

I had tested the behavior of drag drop and resizing extensively before this change and it worked consistently and did not display and performance issues that would indicate needing useCallback.  I would prefer to leave as is unless you see something I may have done incorrectly when adding the useCallback code...  Thanks!

michael-s-molina reviewed (commented):
Thanks for the PR @eric-briscoe! I left some first-pass comments.

eric-briscoe committed:
Update superset-frontend/src/components/DesignSystem.stories.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/DesignSystem.stories.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/index.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/NumericCell/index.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/NumericCell/index.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.test.tsx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/Table.overview.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Update superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.overview.mdx

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

michael-s-molina commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+
+  getColumnIndex = (): number => {
+    let index = -1;
+    const parent: HTMLElement | null | undefined = this.columnRef
+      ?.parentNode as HTMLElement;
```
```suggestion
    const parent = this.columnRef?.parentNode;
```

michael-s-molina commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+  };
+
+  handleColumnDragStart = (ev: DragEvent): void => {
+    const target: IInteractiveColumn = ev?.currentTarget as IInteractiveColumn;
```
```suggestion
    const target = ev?.currentTarget as IInteractiveColumn;
```

michael-s-molina commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+    const data = ev.dataTransfer?.getData?.(SUPERSET_TABLE_COLUMN);
+    if (data) {
+      ev.preventDefault();
+      const parent: Element = (ev.currentTarget as HTMLElement)
```
```suggestion
      const parent = (ev.currentTarget as HTMLElement)
```

eric-briscoe replied:
👍

michael-s-molina commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+  };
+
+  handleMouseDown = (event: MouseEvent) => {
+    const target: IInteractiveColumn =
+      event?.currentTarget as IInteractiveColumn;
```
```suggestion
    const target = event?.currentTarget as IInteractiveColumn;
```

eric-briscoe replied:
👍

michael-s-molina commented on superset-frontend/src/components/Table/utils/InteractiveTableUtils.ts:
```diff
@@ -0,0 +1,237 @@
+
+  handleMouseMove = (event: MouseEvent) => {
+    if (this.resizable === true && !this.isDragging) {
+      const target: IInteractiveColumn =
+        event.currentTarget as IInteractiveColumn;
```
```suggestion
      const target = event.currentTarget as IInteractiveColumn;
```

michael-s-molina commented on superset-frontend/src/components/Table/utils/utils.test.ts:
```diff
@@ -0,0 +1,46 @@
+
+import { withinRange } from './utils';
+
+test('withinRange', () => {
```
Maybe split this into multiple test cases?
- test('ranges', ...)
- test('negative numbers', ...)
- test('invalid input types', ...)

eric-briscoe replied:
👍

michael-s-molina commented on superset-frontend/src/components/Table/utils/utils.ts:
```diff
@@ -0,0 +1,40 @@
+ * @param value number coordinate to determine if it is within bounds of the targetCoordinate - threshold.  Must be positive and less than maximum.
+ * @param maximum number max value for the test range.  Must be positive and greater than value
+ * @param threshold number values to determine a range from maximum - threshold.  Must be positive and greater than zero.
+ * @returns
```
Can you document the return here?

eric-briscoe replied:
👍

eric-briscoe commented on superset-frontend/src/components/Table/Table.stories.tsx:
```diff
@@ -0,0 +1,471 @@
+
+// eslint-disable-next-line no-alert
+const handleClick = (data: object, index: number) =>
+  alert(`I was Clicked: ${JSON.stringify(data)}, index: ${index}`);
```
@lyndsiWilliams Great point, using storybook actions as suggested by @michael-s-molina

eric-briscoe commented on superset-frontend/src/components/Table/Table.stories.tsx:
```diff
@@ -0,0 +1,471 @@
+        index={index}
+        onClick={(row: object, index: number) =>
+          // eslint-disable-next-line no-alert
+          alert(`Cell was clicked: row ${index}, row: ${JSON.stringify(row)}`)
```
using storybook actions as suggested by @michael-s-molina

eric-briscoe commented on superset-frontend/src/components/MetadataBar/MetadataBar.stories.mdx:
```diff
@@ -1,17 +1,25 @@
-# Variations
+## Basic example
+
+<Story id="design-system-components-metadatabar-examples--basic" />
```
👍

eric-briscoe commented on superset-frontend/src/components/Dropdown/index.tsx:
```diff
@@ -66,14 +67,35 @@ const MenuDotsWrapper = styled.div`
+}
 export interface DropdownProps extends DropDownProps {
   overlay: React.ReactElement;
+  orientation?: IconOrientation;
```
👍

eric-briscoe commented on superset-frontend/src/components/Table/Table.stories.tsx:
```diff
@@ -0,0 +1,471 @@
+
+// eslint-disable-next-line no-alert
+const handleClick = (data: object, index: number) =>
+  alert(`I was Clicked: ${JSON.stringify(data)}, index: ${index}`);
```
👍

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.test.tsx:
```diff
@@ -0,0 +1,49 @@
+
+import React from 'react';
+import { render, screen } from 'spec/helpers/testing-library';
+import { configure } from '@testing-library/react';
```
👍

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/index.tsx:
```diff
@@ -0,0 +1,146 @@
+const FILTER = 'drop-shadow(0px 6px 16px rgba(0, 0, 0, 0.08))';
+
+const StyledMenu = styled(Menu)`
+  box-shadow: ${SHADOW} !important;
```
@michael-s-molina antd seems to be dynamically applying css when the menu pops open from the table and overrides any css I set even with specific selectors.  This was the only way I could get the menu to have a shadow.  I could not identify why antd was forcing shadow: none dynamically when menu pops open.  I am open to ideas if you think something else will work, or know the underlying issues stems from

michael-s-molina replied:
👍🏼

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/index.tsx:
```diff
@@ -0,0 +1,146 @@
+const FILTER = 'drop-shadow(0px 6px 16px rgba(0, 0, 0, 0.08))';
+
+const StyledMenu = styled(Menu)`
+  box-shadow: ${SHADOW} !important;
```
@michael-s-molina I am intentionally punting on using theme here as we need to discuss where shadows are defined in theme.  There are shadow definitions in Figma for components like menu, which I extracted and copied here for now.  Once the shadow definitions are define in them we can update this.

michael-s-molina replied:
👍🏼

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ButtonCell/ButtonCell.stories.tsx:
```diff
@@ -0,0 +1,63 @@
+} as ComponentMeta<typeof ButtonCell>;
+
+// eslint-disable-next-line no-alert
+const clickHandler = () => alert(`I was Clicked`);
```
👍

eric-briscoe commented on superset-frontend/src/components/Table/sorters.ts:
```diff
@@ -0,0 +1,36 @@
+ * @returns boolean
+ */
+export const alphabeticalSort = (key: string, a: object, b: object) =>
+  a?.[key]?.localeCompare?.(b?.[key]);
```
👍

eric-briscoe commented on superset-frontend/src/components/Table/sorters.ts:
```diff
@@ -0,0 +1,36 @@
+ * @param b Second row object to compare
+ * @returns boolean
+ */
+export const numericalSort = (key: string, a: object, b: object) =>
```
👍

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ButtonCell/ButtonCell.test.tsx:
```diff
@@ -0,0 +1,43 @@
+
+import React from 'react';
+import { render, screen } from 'spec/helpers/testing-library';
+import { configure } from '@testing-library/react';
```
👍

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.test.tsx:
```diff
@@ -0,0 +1,49 @@
+import { exampleMenuOptions, exampleRow } from './fixtures';
+
+test('renders with default props', async () => {
+  configure({ testIdAttribute: 'data-test' });
```
👍

eric-briscoe committed:
Addresses PR comments

eric-briscoe commented:
> I went through and found some things/had some questions. I also noticed that a storybook log was pushed up to `superset-frontend/build-storybook.log`. I don't see storybook logs elsewhere in the codebase, was this log meant to be part of this PR?

@lyndsiWilliams I removed the build-storybook.log file.  Thanks!

> I went through and found some things/had some questions. I also noticed that a storybook log was pushed up to `superset-frontend/build-storybook.log`. I don't see storybook logs elsewhere in the codebase, was this log meant to be part of this PR?

@lyndsiWilliams I removed that file, thanks for catching that!

eric-briscoe commented on superset-frontend/src/components/Table/cell-renderers/ActionCell/ActionCell.test.tsx:
```diff
@@ -0,0 +1,49 @@
+  expect(clickHandler).toHaveBeenCalled();
+});
+
+test('appendDataToMenu utility', () => {
+  exampleMenuOptions.forEach(item => expect(item?.row).toBeUndefined());
+  const modifiedMenuOptions = appendDataToMenu(exampleMenuOptions, exampleRow);
+  modifiedMenuOptions.forEach(item => expect(item?.row).toBeDefined());
+});
```
@lyndsiWilliams this unit test verifies that the `appendDataToMenu` utility function works as expected.  This is a way of testing code used within components without having to invoke the entire component lifecycle to see if a function behaves correctly.  Do you have any opinions on what could be added to avoid confusion / make this more clear? thanks!

lyndsiWilliams replied:
Oh I see, thank you for the explanation! Maybe just a comment in the test explaining what it's doing just in case 😁

eric-briscoe replied:
👍

eric-briscoe committed:
Apply suggestions from code review

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>

eric-briscoe committed:
Consolidates initializeInteractiveTable function into the useEffect it was called from

eric-briscoe committed:
Improve comments on why derivedColumns is excluded for effect dependencies

michael-s-molina reviewed (approved):
LGTM. Thanks for all the hard work and for addressing the comments @eric-briscoe!

eric-briscoe committed:
Adding comment in test file to clarify appendDataToMenu test purpose based on PR question

eric-briscoe commented:
> LGTM. Thanks for all the hard work and for addressing the comments @eric-briscoe!

Thank you for the thorough reviews @michael-s-molina @lyndsiWilliams @AAfghahi @pkdotson 😄
